### PR TITLE
feat: Creator Rewards

### PR DESCRIPTION
**Pull request by kulkarohan**

### CODE REVIEW

File: src/rewards/RewardsManager.sol [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+/** TODO
+- variants of addReward params and events
+- withdraw w/ recipient
+- withdraw w/ sig?
+ */
+contract RewardsManager is IRewardsManager {
+    mapping(address => uint256) public ethBalance;
+
+    function addReward(address recipient1, uint256 amount1, address recipient2, uint256 amount2, address recipient3, uint256 amount3) external payable {
```
Comment (iainnash) [inline src/rewards/RewardsManager.sol:44]: can we add a 4byte rewardtype here?

`4bytes(keccak256("ZORA_PAID_MINT_REWARDS"))` for instance would be quite helpful for analytics.

File: src/rewards/RewardsManager.sol [hunk 2]
```diff
@@ -0,0 +1,91 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.17;
+
+interface IRewardsManager {
```
Comment (iainnash) [inline src/rewards/RewardsManager.sol:4]: Can this break out to another file?

File: src/rewards/CreatorRewards.sol [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+        if (finder == address(0)) {
+            finder = ZORA_REWARD_RECIPIENT;
+        }
+
+        if (lister == address(0)) {
+            lister = ZORA_REWARD_RECIPIENT;
+        }
+
+        REWARDS_MANAGER.addReward{value: totalReward}(creator, creatorReward, ZORA_REWARD_RECIPIENT, zoraReward, finder, finderReward, lister, listerReward);
```
Comment (oveddan) [inline src/rewards/CreatorRewards.sol:66]: this would be more readable if the REWARDS_MANAGER was camelCase

File: src/rewards/RewardsManager.sol [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    event RewardsAdded(address indexed recipient1, uint256 amount1, address indexed recipient2, uint256 amount2, address indexed recipient3, uint256 amount3);
+    event RewardsAdded(
+        address indexed recipient1,
```
Comment (oveddan) [inline src/rewards/RewardsManager.sol:7]: how about an array here?  and an event with a different name than an overload?

File: test/fee/MintFeeManager.t.sol [hunk 5]
```diff
@@ -11,77 +11,77 @@ import {IZoraCreator1155Factory} from "../../src/interfaces/IZoraCreator1155Fact
...
-        vm.prank(admin);
-        target.addPermission(tokenId, minter, adminRole);
-
-        vm.deal(admin, mintPrice);
-        vm.expectRevert(abi.encodeWithSelector(IMintFeeManager.CannotSendMintFee.selector, _recipient, mintPrice));
-        vm.prank(admin);
-        target.mint{value: mintPrice}(SimpleMinter(payable(minter)), tokenId, quantity, abi.encode(recipient));
-    }
-}
+// contract MintFeeManagerTest is Test {
```
Comment (oveddan) [inline test/fee/MintFeeManager.t.sol:14]: was commenting this out intentional? is what's removed adequately covered in other tests?

File: src/nft/ZoraCreator1155Impl.sol [hunk 6]
```diff
@@ -372,15 +372,35 @@ contract ZoraCreator1155Impl is
...
     /// @param minterArguments The arguments to pass to the minter
-    function mint(IMinter1155 minter, uint256 tokenId, uint256 quantity, bytes calldata minterArguments) external payable nonReentrant {
+    /// @param finder The user or platform that finds the sale
+    /// @param lister The originator of the sale
+    function mint(
+        IMinter1155 minter,
+        uint256 tokenId,
+        uint256 quantity,
+        bytes calldata minterArguments,
+        address finder,
```
Comment (oveddan) [inline src/nft/ZoraCreator1155Impl.sol:382]: I'd be hesitant to force all users of the contract to now pass these arguments additionally if they're only used on a small % of contracts. Adds to gas costs + complexity of the api.  I'd lean to adding another method that allows for these additional arguments.

File: test/minters/fixed-price/ZoraCreatorFixedPriceSaleStrategy.t.sol [hunk 7]
```diff
@@ -10,24 +10,31 @@ import {ICreatorRoyaltiesControl} from "../../../src/interfaces/ICreatorRoyaltie
...
+        rewardsManager = new RewardsManager();
+        ZoraCreator1155Impl targetImpl = new ZoraCreator1155Impl(address(rewardsManager), address(0), address(0));
         Zora1155 proxy = new Zora1155(address(targetImpl));
         target = ZoraCreator1155Impl(address(proxy));
         target.initialize("test", "test", ICreatorRoyaltiesControl.RoyaltyConfiguration(0, 0, address(0)), admin, emptyData);
         fixedPrice = new ZoraCreatorFixedPriceSaleStrategy();
     }
 
+    function computeTotalReward(uint256 numTokens) internal pure returns (uint256) {
+        return numTokens * 0.000999 ether;
```
Comment (oveddan) [inline test/minters/fixed-price/ZoraCreatorFixedPriceSaleStrategy.t.sol:35]: can you pull the 0.000999 ether fee value from the smart contract, so that the logic of the computation (where the value comes from) is clearer?

File: test/minters/fixed-price/ZoraCreatorFixedPriceSaleStrategy.t.sol [hunk 8]
```diff
@@ -323,12 +338,16 @@ contract ZoraCreatorFixedPriceSaleStrategyTest is Test {
...
         vm.expectRevert(abi.encodeWithSignature("WrongValueSent()"));
-        target.mint{value: 0.9 ether}(fixedPrice, newTokenId, 1, abi.encode(tokenRecipient, ""));
+        target.mint{value: 0.9 ether + totalReward}(fixedPrice, newTokenId, 1, abi.encode(tokenRecipient, ""), address(0), address(0));
+
         vm.expectRevert(abi.encodeWithSignature("WrongValueSent()"));
-        target.mint{value: 1.1 ether}(fixedPrice, newTokenId, 1, abi.encode(tokenRecipient, ""));
-        target.mint{value: 1 ether}(fixedPrice, newTokenId, 1, abi.encode(tokenRecipient, ""));
+        target.mint{value: 1.1 ether + totalReward}(fixedPrice, newTokenId, 1, abi.encode(tokenRecipient, ""), address(0), address(0));
+
+        target.mint{value: 1 ether + totalReward}(fixedPrice, newTokenId, 1, abi.encode(tokenRecipient, ""), address(0), address(0));
```
Comment (oveddan) [inline test/minters/fixed-price/ZoraCreatorFixedPriceSaleStrategy.t.sol:350]: all of the changing of the tests here to add empty address fields seems like a bit of an antipattern, like these arguments shouldn't exist on the main `mint` function if they're not always used.

File: src/nft/ZoraCreator1155Impl.sol [hunk 9]
```diff
@@ -372,15 +372,35 @@ contract ZoraCreator1155Impl is
...
+    ) external payable nonReentrant {
         // Require admin from the minter to mint
         _requireAdminOrRole(address(minter), tokenId, PERMISSION_BIT_MINTER);
 
-        // Get value sent and handle mint fee
-        uint256 ethValueSent = _handleFeeAndGetValueSent(quantity);
+        uint256 remainingEth = _parseAttachedEth(msg.value, quantity);
+
+        // If this is a free mint:
+        if (remainingEth == 0) {
```
Comment (oveddan) [inline src/nft/ZoraCreator1155Impl.sol:391]: seems like these are two distinct paths and new features, can you make sure a new test exists for each of them, to demonstrate the new features work properly and their edge cases properly tested?  It would also improve the documentation.

**Comment by oveddan** [comment]
Can you please add to the PR description:
* high level overview of the new features
* description of the logic to calculate the rewards, how/when they are distributed

File: src/rewards/RewardsManager.sol [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+ */
+contract RewardsManager is IRewardsManager {
+    mapping(address => uint256) public ethBalance;
+
+    function addReward(address recipient1, uint256 amount1, address recipient2, uint256 amount2, address recipient3, uint256 amount3) external payable {
+        if (msg.value != (amount1 + amount2 + amount3)) {
+            revert INVALID_AMOUNT();
+        }
+
+        ethBalance[recipient1] += amount1;
```
Comment (oveddan) [inline src/rewards/RewardsManager.sol:49]: there are now three storage writes and adds to track rewards, even in the case there is no finder or lister.   I'd have an overload that just writes to a single ethBalance, for the most common case when there is no finder or lister, and call that when there is no finder or lister.

File: src/rewards/CreatorRewards.sol [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+        if (lister == address(0)) {
+            lister = ZORA_REWARD_RECIPIENT;
+        }
+
+        REWARDS_MANAGER.addReward{value: totalReward}(ZORA_REWARD_RECIPIENT, zoraReward, finder, finderReward, lister, listerReward);
+
+        emit PaidMintRewardsTransferred(finder, finderReward, lister, listerReward);
+    }
+
+    function _computeFreeMintRewards(
```
Comment (oveddan) [inline src/rewards/CreatorRewards.sol:89]: would be good to have an overload that computes this when there is no finder or lister

File: src/nft/ZoraCreator1155Impl.sol [hunk 12]
```diff
@@ -59,7 +59,7 @@ contract ZoraCreator1155Impl is
     /// @notice Factory contract
     IFactoryManagedUpgradeGate internal immutable factory;
 
-    constructor(uint256 _mintFeeAmount, address _mintFeeRecipient, address _factory) MintFeeManager(_mintFeeAmount, _mintFeeRecipient) initializer {
+    constructor(address _rewardsManager, address _zoraRewardRecipient, address _factory) CreatorRewards(_rewardsManager, _zoraRewardRecipient) initializer {
```
Comment (oveddan) [inline src/nft/ZoraCreator1155Impl.sol:62]: we should consider allowing this to be address(0) - and then having it default if that's set to a defaul IRewardsManager, to make it easier to work with
